src/Diagnostics/dump_tendencies.jl

```diff
@@ -0,0 +1,251 @@
+    aux,
+    diffusive,
+    hyperdiffusive,
+    t,
+    direction,
+)
+    _args_fx1 = (; state, aux, t, direction)
+    _args_fx2 = (; state, aux, t, diffusive, hyperdiffusive)
+    _args_src = (; state, aux, t, direction, diffusive)
+
+    cache_fx1 = precompute(bl, _args_fx1, Flux{FirstOrder}())
+    cache_fx2 = precompute(bl, _args_fx2, Flux{SecondOrder}())
+    cache_src = precompute(bl, _args_src, Source())
+
+    args_fx1 = merge(_args_fx1, (; precomputed = cache_fx1))
+    args_fx2 = merge(_args_fx2, (; precomputed = cache_fx2))
+    args_src = merge(_args_src, (; precomputed = cache_src))
+
+    return (args_fx1, args_fx2, args_src)
+end
+
+function dump_tendencies_init(dgngrp, t)
+    dg = Settings.dg
+    Q = Settings.Q
+    FT = eltype(Q)
+    bl = Settings.dg.balance_law
+    mpicomm = Settings.mpicomm
+    mpirank = MPI.Comm_rank(mpicomm)
+
+    if mpirank == 0
+        if array_device(Q) isa CPU
+            prognostic_data = Q.realdata
+            auxiliary_data = dg.state_auxiliary.realdata
+            gradient_flux_data = dg.state_gradient_flux.realdata
+            hyperdiffusive_data = dg.states_higher_order[2].realdata
+        else
+            prognostic_data = Array(Q.realdata)
+            auxiliary_data = Array(dg.state_auxiliary.realdata)
+            gradient_flux_data = Array(dg.state_gradient_flux.realdata)
+            hyperdiffusive_data = Array(dg.states_higher_order[2].realdata)
+        end
+
+        # get dimensions for the interpolated grid
+        dims = dimensions(dgngrp.interpol)
+
+        state, aux, diffusive, hyperdiffusive = (
+            extract_state(dg, prognostic_data, 1, 1, Prognostic()),
+            extract_state(dg, auxiliary_data, 1, 1, Auxiliary()),
+            extract_state(dg, gradient_flux_data, 1, 1, GradientFlux()),
+            extract_state(dg, hyperdiffusive_data, 1, 1, Hyperdiffusive()),
+        )
+        direction = dg.direction
+        args_fx1, args_fx2, args_src = precompute_args(
+            bl,
+            state,
+            aux,
+            diffusive,
+            hyperdiffusive,
+            t,
+            direction,
+        )
+
+        # set up the variables we're going to be writing
+        vars = OrderedDict()
+        for (tend_fun, tend_type, tend_args) in (
+            (flux, Flux{FirstOrder}(), args_fx1),
+            (flux, Flux{SecondOrder}(), args_fx2),
+            (source, Source(), args_src),
+        )
+            for pv in prognostic_vars(bl)
+                progname = prog_name(pv)
+                for tend in eq_tends(pv, bl, tend_type)
+                    flat_vals = flattened_tuple(
+                        FlattenArr(),
+                        tend_fun(tend, bl, tend_args),
+                    )
+                    for i in 1:length(flat_vals)
+                        varname = "$(progname)_$(tend_name(tend))_$(i)"
+                        vars[varname] =
+                            (tuple(collect(keys(dims))...), FT, Dict())
```

Comment:
Should `tuple(collect(keys(dims))...)` be precomputed? Or will Julia optimize this away?

Reply:
Probably can hoist it; doubt Julia will. It's pretty minor though.